### PR TITLE
JENKINS-61278 JCasC support for managing log recorders

### PR DESCRIPTION
**Pull request by timja**
See [JENKINS-61278](https://issues.jenkins-ci.org/browse/JENKINS-61278).

Example configuration (export is fully supported):
```yaml
jenkins:
  log:
    recorders:
    - name: "Azure VM Agent (Auto)"
      loggers:
      - level: "WARNING"
        name: "com.microsoft.azure.vmagent"
```

JCasC demo / integration test:
https://github.com/jenkinsci/configuration-as-code-plugin/pull/1301

### Proposed changelog entries

* Add configuration-as-code support for managing log recorders.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

**Review by jetersen — APPROVED** [review]
Now that's hot! 👏

File: core/src/main/java/hudson/logging/LogRecorderManager.java
Comment (timja) [inline core/src/main/java/hudson/logging/LogRecorderManager.java]: this breaks compatibility for a couple of plugins 😢 
https://github.com/search?q=org%3Ajenkinsci+logRecorders&type=Code

Comment (casz) [reply core/src/main/java/hudson/logging/LogRecorderManager.java]: image

Comment (timja) [reply core/src/main/java/hudson/logging/LogRecorderManager.java]: anyone got any ideas in how to make this compatible =/

I can readResolve to keep populate the field but plugins expect to be able to add more logrecorders by putting to the map

Comment (timja) [reply core/src/main/java/hudson/logging/LogRecorderManager.java]: ah think I have a solution 😄

Comment (timja) [reply core/src/main/java/hudson/logging/LogRecorderManager.java]: Pushed where I got to but for some reason copying the list back to the original map isn't working, if I try lookup the map afterwards it's empty =/

**Comment by timja** [comment]
Got this when installing the azure-vm-agents plugin:
```
Caused by: java.lang.NoSuchFieldError: targets
	at com.microsoft.azure.vmagent.AzureVMCloud$DescriptorImpl.addLogRecorder(AzureVMCloud.java:1053)
	... 13 more
```

Not sure why though as I can compile against this core version (with some properties to ignore enforcer etc) =/

File: core/src/main/java/hudson/logging/LogRecorderManager.java
Comment (casz) [inline core/src/main/java/hudson/logging/LogRecorderManager.java]: Neat!

**Review by jetersen — APPROVED** [review]
I think this is a good compromise

**Comment by timja** [comment]
I think this works, it doesn't handle the save where someone directly adds a logger to the `logRecorders` map and doesn't `save()` but that's easy enough to fix in plugins.

I want to add a couple of tests around the compatibility checks and an integration test in jcasc.

but reviews are welcomed

**Comment by casz** [comment]
I think before merging this adding downstream PRs might be helpful 🤔 
The search does not reveal that many plugins so should not be too bad.

I'll happily help out in the weekend.

**Comment by markyjackson-taulia** [comment]
+1 to the downstream PR's and I could help you if you need as well

**Comment by timja** [comment]
The plugins are:

- mesos-plugin - looks fine
- nodesharing-plugin (needs a call to .save() added for its logrecorder to continue working)
- azure-vm-agents-plugin - looks fine
- support-core-plugin  - looks fine

I've got an integration test still needed on jcasc side but if anyone wants to add other downstream PRs go for it.

File: core/src/main/java/hudson/logging/LogRecorder.java
Comment (res0nance) [inline core/src/main/java/hudson/logging/LogRecorder.java]: How breaking is this change? If it is fairly breaking should we consider making this private and adding the appropriate public api for it?

Comment (timja) [reply core/src/main/java/hudson/logging/LogRecorder.java]: shouldn't be breaking at all

Comment (oleg-nenashev) [reply core/src/main/java/hudson/logging/LogRecorder.java]: Formally it breaks the binary compatibility. https://wiki.eclipse.org/Evolving_Java-based_APIs_2 . I would prefer to avoid this change if possible, especially since it also creates concurrency issues (the public collection is writable from API...)

We could consider doing a breaking change, but in such case I strongly suggest creating a new field and implementing a `readResolve()` migration logic. The new field should be private also

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
It would be great to deliver this feature, but we need to define our strategy w.r.t the breaking change first

File: core/src/main/java/hudson/logging/LogRecorder.java
Comment (timja) [reply core/src/main/java/hudson/logging/LogRecorder.java]: > I would prefer to avoid this change if possible, especially since it also creates concurrency issues (the public collection is writable from API...)

CopyOnWriteList isn't supported by JCasC, would be surprised if this was the first time that this is been an issue.

I can do that but are you sure it's needed for this case?
and any thoughts on a name for the new var? as the one that makes sense is already there

Comment (oleg-nenashev) [reply core/src/main/java/hudson/logging/LogRecorder.java]: `logTargets` 🙈 ?

Comment (timja) [reply core/src/main/java/hudson/logging/LogRecorder.java]: I've made the suggested changes

**Review by oleg-nenashev — DISMISSED** [review]
IIRC XStream does not consult with default constructors when deserializing fields. I suggest to explicitly initialize `loggers` in readResolve()

**Comment by jetersen** [comment]
@oleg-nenashev GitHub supports multiline suggestion 👍

https://github.blog/changelog/2020-02-26-multi-line-code-suggestions-beta/

File: core/src/main/java/hudson/logging/LogRecorder.java
Comment (jetersen) [inline core/src/main/java/hudson/logging/LogRecorder.java:106]: the same is true for targets, this could also be null and would cause a null point exception.

```suggestion
        if (targets != null && !targets.isEmpty()) {
```

**Review by jonbrohauge — COMMENTED** [review]
The textual changes are absolutely for the better. They definitely make more sense now.
I'm not into the code enough to comment on that sensibly.

**Comment by jetersen** [comment]
@oleg-nenashev this no longer has breaking changes.
This would be really great to finally land.

**Comment by alecharp** [comment]
@timja could you have a look at the build failure and fix the PR? thanks!

**Comment by varyvol** [comment]
@oleg-nenashev it also seems like your requests were made. Could you have another look?

**Comment by timja** [comment]
Seems to compile locally just fine after merging in master =/.

Let's see what CI thinks of it

File: core/src/main/java/hudson/logging/LogRecorder.java
Comment (varyvol) [reply core/src/main/java/hudson/logging/LogRecorder.java:110]: @timja did you review this comment? It seems like a valid concern.

**Comment by jetersen** [comment]
Birthday is only 1 and ½ month away 😅
Would be great to land this.

File: core/src/main/java/hudson/logging/LogRecorder.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/logging/LogRecorder.java:4]:
```suggestion
 * Copyright (c) 2004-2021, Sun Microsystems, Inc., Kohsuke Kawaguchi, and other contributors
```

**Comment by jonbrohauge** [comment]
@oleg-nenashev  👍 for your update to the copyright statement.

**Comment by oleg-nenashev** [comment]
@jonbrohauge it was rather to poke the Pipeline which has been already retriggered.

**Comment by jonbrohauge** [comment]
@oleg-nenashev I know.
Just getting back to the hang of things after being "out of office" in respect to Jenkins and the Open-Source community in general.

**Comment by oleg-nenashev** [comment]
Poked the build again (and fixed the merge conflicts)

**Comment by timja** [comment]
Resolved conflicts for about the ~5th time.

and updated the tests to latest incremental:
https://github.com/jenkinsci/configuration-as-code-plugin/pull/1301

Can someone please take a look at this @jenkinsci/core-pr-reviewers

File: core/src/main/java/hudson/logging/LogRecorderManager.java
Comment (res0nance) [inline core/src/main/java/hudson/logging/LogRecorderManager.java:154]: Do we not have to sync the legacy field here?

Comment (timja) [reply core/src/main/java/hudson/logging/LogRecorderManager.java:154]: generally this is just to maintain API compatibility, if this does or doesn't it won't affect users.

I plan to file PRs to update consumers to the new API anyway.

File: core/src/main/java/hudson/logging/LogRecorder.java
Comment (basil) [inline core/src/main/java/hudson/logging/LogRecorder.java:136]: My strong preference would be to add a test for this `readResolve()` method using `@LocalData` as in e.g. https://github.com/jenkinsci/email-ext-plugin/blob/master/src/test/java/hudson/plugins/emailext/ExtendedEmailPublisherDescriptorTest.java#L512-L704

**Review by basil — APPROVED** [review]

**Comment by timja** [comment]
Thanks @basil, I plan to prepare PRs today / tomorrow to upgrade consuming plugins instead, I would rather spend time removing the old API.

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by basil** [comment]
> I plan to prepare PRs today / tomorrow to upgrade consuming plugins instead, I would rather spend time removing the old API.

But [now](https://github.com/jenkinsci/periodicbackup-plugin/blob/45c844327210ace5cedd62e10ded7c98c6416757/src/main/java/org/jenkinsci/plugins/periodicbackup/PeriodicBackupPlugin.java#L61=) [four](https://github.com/jenkinsci/azure-vm-agents-plugin/blob/0a97a847315ad69de4025442dfa050891d1520c0/src/main/java/com/microsoft/azure/vmagent/AzureVMCloud.java#L1016=) [months](https://github.com/jenkinsci/mesos-plugin/blob/7da7bc60883eec7fbb8e1fe888500d735bd297ba/src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java#L386-L389=) [later](https://github.com/jenkinsci/servicenow-devops-plugin/blob/100987b498e63dcea22e0d49d942c8262092c302/src/main/java/io/jenkins/plugins/utils/GenericUtils.java#L126-L130=)?

**Comment by basil** [comment]
> > I plan to prepare PRs today / tomorrow to upgrade consuming plugins instead, I would rather spend time removing the old API.
> 
> But [now](https://github.com/jenkinsci/periodicbackup-plugin/blob/45c844327210ace5cedd62e10ded7c98c6416757/src/main/java/org/jenkinsci/plugins/periodicbackup/PeriodicBackupPlugin.java#L61=) [four](https://github.com/jenkinsci/azure-vm-agents-plugin/blob/0a97a847315ad69de4025442dfa050891d1520c0/src/main/java/com/microsoft/azure/vmagent/AzureVMCloud.java#L1016=) [months](https://github.com/jenkinsci/mesos-plugin/blob/7da7bc60883eec7fbb8e1fe888500d735bd297ba/src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java#L386-L389=) [later](https://github.com/jenkinsci/servicenow-devops-plugin/blob/100987b498e63dcea22e0d49d942c8262092c302/src/main/java/io/jenkins/plugins/utils/GenericUtils.java#L126-L130=)?

[JENKINS-68752](https://issues.jenkins.io/browse/JENKINS-68752)

**Comment by timja** [comment]
> > I plan to prepare PRs today / tomorrow to upgrade consuming plugins instead, I would rather spend time removing the old API.
> 
> But [now](https://github.com/jenkinsci/periodicbackup-plugin/blob/45c844327210ace5cedd62e10ded7c98c6416757/src/main/java/org/jenkinsci/plugins/periodicbackup/PeriodicBackupPlugin.java#L61=) [four](https://github.com/jenkinsci/azure-vm-agents-plugin/blob/0a97a847315ad69de4025442dfa050891d1520c0/src/main/java/com/microsoft/azure/vmagent/AzureVMCloud.java#L1016=) [months](https://github.com/jenkinsci/mesos-plugin/blob/7da7bc60883eec7fbb8e1fe888500d735bd297ba/src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java#L386-L389=) [later](https://github.com/jenkinsci/servicenow-devops-plugin/blob/100987b498e63dcea22e0d49d942c8262092c302/src/main/java/io/jenkins/plugins/utils/GenericUtils.java#L126-L130=)?

https://github.com/jenkinsci/jenkins/pull/10612

**Pull request merged**
